qlib/Mapper.qm

```diff
@@ -947,28 +969,29 @@ Mapper mapv(DataMap);
             if (exists m.constant)
                 v = m.constant;
             else if (exists m.index) {
-                v = do_list ? (map m.index + count + $#, v) : m.index + count;
+                v = do_list ? (map m.index + count + $#, v) : m.index + count; # FIXME iterating empty list
             }
             else if (exists m.runtime) {
                 any rtv = m_runtime{m.runtime};
-                v = do_list ? (map rtv, v) : rtv;
+                v = do_list ? (map rtv, v) : rtv; # FIXME iterating empty list
             }
             else if (m.struct) {
+                # NOTE: hash of lists not supported with struct # FIXME do we ensure it?
                 v = rec;
                 map v = v{m.struct[$1]}, xrange(0, m.struct.size() - 1);
             }
             else
-                v = rec{name} ?? NOTHING;
+                v = rec{name} ?? NOTHING; # mitigate NULL -> NOTHING (?)
 
             if (do_list) {
                 if (!exists v)
-                    v = map NOTHING, xrange(do_list - 1);
+                    v = list_size > 0 ? (map NOTHING, xrange(list_size - 1)) : list();
```

Comment:
ok, using () instead of list(). Otherwise I don't agree. this code IS processed when you pass hash of empty lists (there is a test for it in this pull request already), and xrange(-1) returns two elements, which is IMHO not correct in this case.